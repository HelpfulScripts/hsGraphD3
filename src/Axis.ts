/**
 * # Axis class
 * 
 */

/** */
import { axisTop }          from 'd3';
import { axisRight }        from 'd3';
import { log as gLog }      from 'hsutil';   const log = gLog('Axis');
import { GraphComponent }   from './GraphComponent'; 
import { GraphCfg }         from './GraphComponent';
import { ComponentDefaults }from './GraphComponent';
import * as def             from './Defaults';
import { ScaleDefaults }    from './Scale';
import { UnitVp, d3Base }   from './Defaults';


export enum Direction {
    horizontal  = 'hor',
    vertical    = 'ver'
}

export interface AxisDefaults extends ComponentDefaults {
    color:      string;
    line:       def.Line;
    tickWidth:  UnitVp;
    tickMarks:  def.Line;
    tickLabel:  def.TextStyle;
}

export interface AxesDefaults extends ComponentDefaults {
    color:      string;
    hor:        AxisDefaults;
    ver:        AxisDefaults;
} 


export class Axes extends GraphComponent {
    private axes: Axis[] = [];

    constructor(cfg:GraphCfg) {
        super(cfg, cfg.baseSVG.select('.axes'));
        let axis;
        this.axes.push(new Axis(cfg, Direction.horizontal));
        this.axes.push(new Axis(cfg, Direction.vertical));
    }

    public get componentType() { return 'axes'; }

    public createDefaults():AxesDefaults {
        return {
            color:  '#000',
            hor:    this.axes[0].createDefaults(),
            ver:    this.axes[1].createDefaults()
        };
    }
    
    public renderComponent() {
        this.axes.forEach(axis => axis.renderComponent());
    }
}


export class Axis {
    private dir: Direction;
    private cfg: GraphCfg;
    private svg: d3Base;

    constructor(cfg:GraphCfg, dir:Direction) {
        this.cfg = cfg;
        this.svg = cfg.baseSVG.select('.axes').append('g');
        this.dir = dir;
        this.svg.classed(`${dir}Axis`, true);
    }

    public createDefaults() {
        return {
            color:      'currentColor',
            line:       def.defaultLine(2),
            tickWidth:  10,
            tickMarks:  def.defaultLine(2),
            tickLabel:  def.defaultText()
        };
    }
    
    public renderComponent() {
        const horScales = this.cfg.scales.hor;
        const verScales = this.cfg.scales.ver;
        const style = this.cfg.defaults.axes[this.dir];
        let axis;
        const margins = (<ScaleDefaults>this.cfg.defaults.scales).margin;
        this.cfg.baseSVG.select('.axes')
            .attr('color', this.cfg.defaults.axes.color);
        this.svg
            .attr('stroke', style.line.color)
            .attr('stroke-width', style.line.width)
            .attr('stroke-opacity', style.line.opacity);

        if (this.dir===Direction.horizontal) {
            axis = axisTop(horScales);
            const yCrossing = Math.max(margins.left, Math.min(verScales(0), this.cfg.viewPort.height-margins.right));
            this.svg.attr("transform", `translate(0, ${yCrossing})`);
        } else {
            axis = axisRight(verScales);
            const xCrossing = Math.max(margins.top, Math.min(horScales(0), this.cfg.viewPort.width-margins.bottom));
            this.svg.attr("transform", `translate(${xCrossing}, 0)`);
        }
        axis.tickSize(style.tickWidth);
        this.svg
            .attr('color', style.color)
            .style('font-family',  style.tickLabel.font.family)
            .style('font-size',   `${style.tickLabel.font.size}px`)
            .style('font-style',  style.tickLabel.font.style)
            .style('font-weight', style.tickLabel.font.weight);
        this.svg.call(axis);
        this.svg.selectAll('text')
            .transition().duration(1000)
            .attr('stroke', style.tickLabel.color || 'currentColor')
            .attr('stroke-width', 1);
        }
}