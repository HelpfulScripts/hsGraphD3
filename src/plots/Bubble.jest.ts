import * as hsGraphD3   from '../';
import { log }          from 'hsnode';
import * as d3 from 'd3';
import { DataSet }      from '../Graph';
import { AxesDefaults } from '../Axis';

const root = window.document.createElement("div");
root.style.width = "300px";
root.style.height = "300px";

d3.selection.prototype.transition   = function(){ return this; };
d3.selection.prototype.duration     = function(){ return this; };
d3.selection.prototype.ease         = function(){ return this; };


const data:DataSet = {
    colNames: 
        ['xval', 'yval', 'rval'],
    rows : [
        [-1,      15,     15],
        [0.2,      2,     23],
        [3,       28,     30],
        [7.5,     13,      8]
    ]
};

function createGraph(root:any) {
    const graph = new hsGraphD3.GraphCartesian(root);
    graph.isRendered = () => true;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).hor.numTicksMajor = 2;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMinor = 10;
    (<AxesDefaults>graph.defaults.axes).ver.numTicksMajor = 2;
    return graph;
}

describe('Bubble', () => {
    let graph:hsGraphD3.GraphCartesian;
    beforeAll(() => {
        graph = createGraph(root);
        graph.addSeries('bubble', {x:'xval', y:'yval', r:'rval'});
        graph.render(data);
    });
    it('should plot series', () =>
        expect(root).toMatchSnapshot()
    );
});
