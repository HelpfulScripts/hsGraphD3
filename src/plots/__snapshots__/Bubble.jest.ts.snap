// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bubble should plot series 1`] = `
<div
  style="width: 300px; height: 300px;"
>
  <svg
    class="baseSVG"
    height="100%"
    preserveAspectRatio="xMinYMin meet"
    viewBox="0 0 1000 700"
    width="100%"
  >
    <g
      class="canvas"
    >
      <rect
        class="graphArea"
        fill="#fff"
        fill-opacity="1"
        height="700"
        rx="0"
        ry="0"
        stroke="#00c"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        width="1000"
      />
    </g>
    <g
      class="grids"
    >
      <g
        class="hor-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="470"
          y2="470"
        />
        <line
          x1="20"
          x2="980"
          y1="220"
          y2="220"
        />
      </g>
      <g
        class="hor-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="980"
          y1="670"
          y2="670"
        />
        <line
          x1="20"
          x2="980"
          y1="620"
          y2="620"
        />
        <line
          x1="20"
          x2="980"
          y1="570"
          y2="570"
        />
        <line
          x1="20"
          x2="980"
          y1="520"
          y2="520"
        />
        <line
          x1="20"
          x2="980"
          y1="470"
          y2="470"
        />
        <line
          x1="20"
          x2="980"
          y1="420"
          y2="420"
        />
        <line
          x1="20"
          x2="980"
          y1="370"
          y2="370"
        />
        <line
          x1="20"
          x2="980"
          y1="320"
          y2="320"
        />
        <line
          x1="20"
          x2="980"
          y1="270"
          y2="270"
        />
        <line
          x1="20"
          x2="980"
          y1="220"
          y2="220"
        />
        <line
          x1="20"
          x2="980"
          y1="170"
          y2="170"
        />
        <line
          x1="20"
          x2="980"
          y1="120"
          y2="120"
        />
        <line
          x1="20"
          x2="980"
          y1="70"
          y2="70"
        />
        <line
          x1="20"
          x2="980"
          y1="20"
          y2="20"
        />
      </g>
      <g
        class="ver-major-Grid"
        fill-opacity="0"
        stroke="#444"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="133"
          x2="133"
          y1="670"
          y2="20"
        />
        <line
          x1="698"
          x2="698"
          y1="670"
          y2="20"
        />
      </g>
      <g
        class="ver-minor-Grid"
        fill-opacity="0"
        stroke="#eee"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
      >
        <line
          x1="20"
          x2="20"
          y1="670"
          y2="20"
        />
        <line
          x1="133"
          x2="133"
          y1="670"
          y2="20"
        />
        <line
          x1="246"
          x2="246"
          y1="670"
          y2="20"
        />
        <line
          x1="359"
          x2="359"
          y1="670"
          y2="20"
        />
        <line
          x1="472"
          x2="472"
          y1="670"
          y2="20"
        />
        <line
          x1="585"
          x2="585"
          y1="670"
          y2="20"
        />
        <line
          x1="698"
          x2="698"
          y1="670"
          y2="20"
        />
        <line
          x1="811"
          x2="811"
          y1="670"
          y2="20"
        />
        <line
          x1="924"
          x2="924"
          y1="670"
          y2="20"
        />
      </g>
    </g>
    <g
      class="axes"
      color="currentColor"
    >
      <g
        class="horAxis"
        color="currentColor"
        fill="currentColor"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="16px"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="middle"
        transform="translate(0, 670)"
      >
        <path
          class="domain"
          d="M20.5,5V0.5H980.5V5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(20.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            -1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(133.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            0
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(246.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            1
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(359.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(472.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            3
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(585.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(698.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            5
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(811.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(924.5,0)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            y2="5"
          />
          <text
            dy="0.71em"
            fill="currentColor"
            y="8"
          >
            7
          </text>
        </g>
      </g>
      <g
        class="verAxis"
        color="currentColor"
        fill="currentColor"
        fill-opacity="0"
        font-family="sans-serif"
        font-size="16px"
        font-style="normal"
        font-weight="100"
        stroke="currentColor"
        stroke-dasharray="none"
        stroke-opacity="1"
        stroke-width="1"
        text-anchor="end"
        transform="translate(133, 0)"
      >
        <path
          class="domain"
          d="M-5,670.5H0.5V20.5H-5"
          stroke="currentColor"
        />
        <g
          class="tick"
          opacity="1"
          transform="translate(0,670.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            2
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,620.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            4
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,570.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            6
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,520.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            8
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,470.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            10
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,420.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            12
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,370.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            14
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,320.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            16
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,270.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            18
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,220.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            20
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,170.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            22
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,120.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            24
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,70.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            26
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,20.5)"
        >
          <line
            fill-opacity="0"
            stroke="currentColor"
            stroke-dasharray="none"
            stroke-opacity="1"
            stroke-width="2"
            x2="-5"
          />
          <text
            dy="0.32em"
            fill="currentColor"
            x="-8"
          >
            28
          </text>
        </g>
      </g>
    </g>
    <g
      class="title"
    >
      <text />
    </g>
    <g
      class="series"
    >
      <g
        class="series0"
        style="color: rgb(27, 158, 119);"
      >
        <g
          class="markers"
          fill="currentColor"
          fill-opacity="0.75"
          stroke="currentColor"
          stroke-dasharray="none"
          stroke-opacity="1"
          stroke-width="1"
        >
          <circle
            class="popup"
            cx="20"
            cy="345"
            r="10"
          />
          <circle
            class="popup"
            cx="156"
            cy="670"
            r="15"
          />
          <circle
            class="popup"
            cx="472"
            cy="20"
            r="20"
          />
          <circle
            class="popup"
            cx="980"
            cy="395"
            r="5"
          />
        </g>
      </g>
    </g>
    <g
      class="popup"
    />
    <style>
      .popup:hover { filter: brightness(0.8);}
    </style>
  </svg>
</div>
`;
